### PR TITLE
Add DiffFormatter for the diff of revisions in two repositories

### PR DESCRIPTION
**Pull request by minwoox**
Motivation:
To implement #575, we should be able to make diffs between two different repositories.

Modifications:
- Add `TwoRepositoriesDiffFormatter` and `TwoRepositoriesTreeWalk` to make diffs between two different repositories.
- Misc
  - Use `@TempDir` in `GitRepositoryManagerTest` so that the temporarily created directories for the test are not left anymore.

Result:
- Ready to implement #575

### CODE REVIEW

**Comment by KarboniteKream** [comment]
Looks like `@TempDir` still has trouble deleting the directory on Windows 😭 

It seems to be due to having read-only files in the directory, but AFAIK the extension should try to make them readable before deletion. I wonder if the issue is related to JUnit or JGit...

**Comment by minwoox** [comment]
> When does it leave the files? I've tried replicating it (on MacOS), but TemporaryFolderExtension always deletes the files after itself 🤔

I think https://github.com/line/centraldogma/blob/master/server/src/test/java/com/linecorp/centraldogma/server/internal/storage/repository/git/GitRepositoryManagerTest.java#L124
this line creates a directory and does not remove it. 😅 

> Looks like @TempDir still has trouble deleting the directory on Windows 😭

Ah, I didn't know about it. 😅 

> Perhaps GitRepositoryManager is still holding some file locks?

I guess so, let me see what happened. Thanks a lot for your investigation. 😄

**Comment by KarboniteKream** [comment]
> Ah, I didn't know about it. 😅

It's possible it's due to the test implementation. When I was implementing JUnit 5 support, I wasn't too familiar with how Central Dogma works, so I wasn't able to figure it out. If your latest commit works, I'll check other tests and cleanup `TemporaryFolderExtension`.

**Comment by minwoox** [comment]
> It's possible it's due to the test implementation. When I was implementing JUnit 5 support, I wasn't too familiar with how Central Dogma works, so I wasn't able to figure it out. If your latest commit works, I'll check other tests and cleanup TemporaryFolderExtension.

I guess it worked at this time. 😄  Oh, could you? Thanks a lot for taking care of this. 🙇‍♂️

File: server/src/main/java/com/linecorp/centraldogma/server/internal/storage/repository/git/TwoRepositoriesTreeWalk.java [hunk 1]
```diff
@@ -0,0 +1,545 @@
...
+            }
+            if (trace[1] == -1) {
+                throw new ArrayIndexOutOfBoundsException(delta);
+            }
+            prevPtr = trace[0];
+            currPtr = trace[1];
+            parseEntry();
+        }
+
+        private void parseEntry() {
```
Comment (ikhoon) [inline server/src/main/java/com/linecorp/centraldogma/server/internal/storage/repository/git/TwoRepositoriesTreeWalk.java:440]: Could you leave an entry format of a tree as a comment?
```
[mode] [file/folder name]\0[SHA-1 of referencing blob or tree]
```

**Review by ikhoon — COMMENTED** [review]
Mostly looking good. I've learned a lot of Git internals.

File: server/src/main/java/com/linecorp/centraldogma/server/internal/storage/repository/git/TwoRepositoriesDiffFormatter.java [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+import org.eclipse.jgit.diff.DiffEntry;
+import org.eclipse.jgit.diff.DiffEntry.ChangeType;
+import org.eclipse.jgit.lib.AbbreviatedObjectId;
+import org.eclipse.jgit.lib.FileMode;
+import org.eclipse.jgit.lib.MutableObjectId;
+import org.eclipse.jgit.lib.ObjectId;
+
+import com.google.common.collect.ImmutableList;
+
+final class TwoRepositoriesDiffFormatter {
```
Comment (trustin) [inline server/src/main/java/com/linecorp/centraldogma/server/internal/storage/repository/git/TwoRepositoriesDiffFormatter.java:41]: `InterRepositoryDiffFormatter` or `CrossRepositoryDiffFormatter` ?

File: server/src/main/java/com/linecorp/centraldogma/server/internal/storage/repository/git/TwoRepositoriesTreeWalk.java [hunk 3]
```diff
@@ -0,0 +1,545 @@
...
+import org.eclipse.jgit.lib.ObjectId;
+import org.eclipse.jgit.lib.ObjectReader;
+import org.eclipse.jgit.lib.Repository;
+import org.eclipse.jgit.revwalk.RevTree;
+import org.eclipse.jgit.revwalk.RevWalk;
+import org.eclipse.jgit.treewalk.AbstractTreeIterator;
+import org.eclipse.jgit.treewalk.filter.TreeFilter;
+import org.eclipse.jgit.util.RawParseUtils;
+
+final class TwoRepositoriesTreeWalk implements AutoCloseable {
```
Comment (trustin) [inline server/src/main/java/com/linecorp/centraldogma/server/internal/storage/repository/git/TwoRepositoriesTreeWalk.java:52]: `InterRepositoryTreeWalk` or `CrossRepositoryTreeWalk` ?

File: server/src/test/java/com/linecorp/centraldogma/server/internal/storage/repository/git/DiffFormatterTest.java [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+        //  The revision number - 1 of Single repo is the corresponding revision of New repo.
+
+        //  New repo:
+        //  Revision     1            2                          3              4           5             6
+        //  initial commit  Add /a.json              Modify /a.json                           Rename to /aa.json
+        //  initial commit  Add /a.txt
+        //  initial commit  Add /b/b.json
+        //  initial commit  Add /b/b.txt                           Remove /b/b.txt
+        //  initial commit                                                         Add /c.json
+        //  initial commit                                                         Add /c.txt
```
Comment (trustin) [inline server/src/test/java/com/linecorp/centraldogma/server/internal/storage/repository/git/DiffFormatterTest.java:108]: How about using some vertical lines (`|`) to make it clear which change belongs to which commit?

**Review by trustin — COMMENTED** [review]
I didn't check the implementation detail assuming that jGit folks did a good job already and no big changes were made when forking it. Let me know if you need more detailed review. :smile:

**Comment by minwoox** [comment]
> I didn't check the implementation detail assuming that jGit folks did a good job already and no big changes were made when forking it.

Thanks for the review and no problem at all. The point that I wanted to get reviewed was whether I missed some test cases or not. 😄

File: server/src/main/java/com/linecorp/centraldogma/server/internal/storage/repository/git/TwoRepositoriesTreeWalk.java [hunk 5]
```diff
@@ -0,0 +1,545 @@
...
+        /**
+         * Offset one past the current entry (first byte of next entry).
+         */
+        private int nextPtr;
+
+        private final ObjectReader reader;
+        @Nullable
+        CanonicalTreeParser matches;
+
+        CanonicalTreeParser(CanonicalTreeParser parent, ObjectReader reader) {
```
[the remark this answers is not in the document]
Comment (minwoox) [reply server/src/main/java/com/linecorp/centraldogma/server/internal/storage/repository/git/TwoRepositoriesTreeWalk.java:290]: Can I just use as it is? 🤔 
I still have to implement the method `createSubtreeIterator` by myself and also I have to override all methods in `AbstractTreeIterator` to delegate the call which is not necessary when just extending `AbstractTreeIterator`. 😄

[the remark this answers is not in the document]
Comment (minwoox) [reply server/src/main/java/com/linecorp/centraldogma/server/internal/storage/repository/git/TwoRepositoriesTreeWalk.java]: @ikhoon Do you have any concerns about it? 🤔

[the remark this answers is not in the document]
Comment (ikhoon) [reply server/src/main/java/com/linecorp/centraldogma/server/internal/storage/repository/git/TwoRepositoriesTreeWalk.java]: Sorry, I missed your comment. I didn't deeply think about the details. Anyway, I'm also fine with the current implementation.

**Review by ikhoon — APPROVED** [review]
Nice work! @minwoox 👍 and sorry for the late reivew.

**Comment by minwoox** [comment]
No problem at all. Thanks a lot for reviewing. 😉

**Comment by minwoox** [comment]
We do not need this diff formatter because we are going to implement the new repository in a way that the newly-created subrepository contains at least N commits that it needs. 
So let me close this. 😄

**Pull request closed without merging**